main.py

```diff
@@ -239,36 +239,39 @@ async def wrapper(message: discord.message):
-    if message.author.name == 'TestBot':
-        message.author.bot = False
-
-    message_copy = MessageCopy(content=message.content, display_name=message.author.display_name, avatar=message.author.display_avatar, attachments=message.attachments, reactions=message.reactions)
-
-    # handle DMs
-    if isinstance(message.channel, discord.DMChannel):
-        LOGGER.info("Beginning DM listener stack execution.")
-        for listener in direct_message_listeners:
-            if await listener(message, message_copy):
-                return
-        await bot.process_commands(message)
-        return
-
-    LOGGER.info("Beginning message listener stack execution.")
-    # use the listeners for bot messages or user messages
-    applicable_listeners = bot_message_listeners if message.author.bot else message_listeners
-    for listener in applicable_listeners:
-        LOGGER.info(f"Executing: {listener}")
-        if await listener(message, message_copy):  # Return early if any listeners return true.
-            return
-    LOGGER.info("End of message listener stack.")
+    with LogContext('on_message(from=' + message.author.name + ', content=' + message.content + ')'):
+        # Don't ignore messages from the testing bot.
+        # Usually process_commands() will ignore messages from bots.
+        if getattr(message.author, 'guild', None) is not None and has_role(message.author, TEST_BOT):
+            message.author.bot = False
+
+        message_copy = MessageCopy(content=message.content, display_name=message.author.display_name, avatar=message.author.display_avatar, attachments=message.attachments, reactions=message.reactions)
+        error = None
+
+        # Select the message listeners to execute.
+        if isinstance(message.channel, discord.DMChannel):
+            listeners = direct_message_listeners
+        else:
+            listeners = bot_message_listeners if message.author.bot else message_listeners
 
-    LOGGER.info("Checking for need to webhook.")
-    await webhook.webhook_if_message_altered(message, message_copy)
+        for listener in listeners:
+            with LogContext(listener.__name__):
+                try:
+                    if await listener(message, message_copy):
+                        return
+                except CommandError as err:
+                    error = err
 
-    LOGGER.info("Processing additional commands.")
-    await bot.process_commands(message)
+        # Replace the original message if it was altered by a listener.
+        if not isinstance(message.channel, discord.DMChannel):
+            await webhook.webhook_if_message_altered(message, message_copy)
+
+        # Report the error after the webhook so that the error message always follows the command,
+        # even if the original command message was deleted and replaced by the webhook.
+        if error:
+            await report_error(message.channel, str(error))
```

Comment:
I'm not sure if this will necessarily work. In certain situations, where the AI creates multiple messages back to back, they will appear in unexpected orders. I think this has to do with the asynchronicity (is that how you spell that?) of the whole API. Interactions with the discord API are not guaranteed to take effect in a specific order.

But no need to worry about that too much here. That is a general issue we may want to tackle separately later.

Reply:
The issue here is that you should display the error message after running the webhook, otherwise the messages appear in the wrong order:

1) User sends command to be executed by a listener, eg a storage command: 1234 :: 0006 :: 1 :: Store the Hive Mxtress
2) The bot sends an error message: "You cannot store the Hive Mxtress"
3) The bot's identity enforcement listener runs.  This deletes message 1 and sends it again, with the default drone avatar and such.

At this point, the chat history looks like:

(Deleted message from User: 1234 :: 0006 :: 1 :: Store the Hive Mxtress)
Bot: "You cannot store the Hive Mxtress"
[Bot acting as user]: 1234 :: 0006 :: 1 :: Store the Hive Mxtress

The command is displayed after the error message, which is confusing.
The code above stores the error message instead of outputting it in step 2, and outputs it after step 3.  This solves the issue.

From looking at the pycord source, it looks like when sending a message it waits for a response from the server before continuing, so messages should appear in the order sent.

Reply:
oh you are probably right. I was thinking about issues we have in different places were our message processing is in a slightly confusing order. In storage it's very noticeable. So this here is probably completely fine.